src/server/actor/runtime/mod.rs

```diff
@@ -0,0 +1,332 @@
+    /// .await;
+    /// # }
+    /// ```
+    pub fn register_method<T>(
+        mut self,
+        method_name: &str,
+        handler: impl Handler<T, ActorState> + Send + Sync,
+    ) -> Self
+    where
+        T: 'static,
+    {
+        let actor_type = self.name.clone();
+        let method_path = format!("/actors/{}/:actor_id/method/{}", actor_type, method_name);
+
+        let reg_func = move |router: Router, runtime: Arc<ActorRuntime>| {
+            router.route(
+                &method_path,
+                put(handler).with_state(ActorState {
+                    actor_type,
+                    runtime,
+                }),
+            )
+        };
+
+        self.method_registrations
+            .insert(method_name.to_string(), Box::new(reg_func));
+        self
+    }
+
+    fn create_actor(&self, actor_id: &str, client: TonicClient) -> Arc<dyn Actor> {
+        let client = ActorContextClient::new(client, &self.name, actor_id);
+
+        (self.factory)(&self.name, actor_id, client) as _
+    }
+}
+
+type ActiveActorMap = Arc<RwLock<HashMap<(String, String), Arc<dyn Actor>>>>;
+type ActorRegistrationMap = Arc<RwLock<HashMap<String, ActorTypeRegistration>>>;
+
+pub struct ActorRuntime {
+    dapr_client: TonicClient,
+
+    registered_actors_types: ActorRegistrationMap,
+    active_actors: ActiveActorMap,
+}
+
+impl ActorRuntime {
+    pub fn new(dapr_client: TonicClient) -> Self {
+        ActorRuntime {
+            dapr_client,
+            registered_actors_types: Arc::new(RwLock::new(HashMap::new())),
+            active_actors: Arc::new(RwLock::new(HashMap::new())),
+        }
+    }
+
+    /// Registers an actor type to be exposed to actor clients.
+    /// # Arguments:
+    /// * `registration` - The [ActorTypeRegistration] that describes the actor implementation.
+    pub async fn register_actor(&self, registration: ActorTypeRegistration) {
+        let name = registration.name.clone();
+        let mut g = self.registered_actors_types.write().await;
+        g.insert(name.clone(), registration);
+        log::info!("registered actor {}", name);
+    }
+
+    pub async fn configure_method_routes(
+        &self,
+        router: Router,
+        runtime: Arc<ActorRuntime>,
+    ) -> Router {
+        let mut router = router;
+        let mut types = self.registered_actors_types.write().await;
+
+        for (_, registration) in types.iter_mut() {
+            for (_, reg_func) in registration.method_registrations.drain() {
+                router = reg_func(router, runtime.clone());
+            }
+        }
+        router
+    }
+
+    pub async fn deactivate_actor(&self, name: &str, id: &str) -> Result<(), ActorError> {
+        let mut actors = self.active_actors.write().await;
+
+        let actor = match actors.remove(&(name.to_string(), id.to_string())) {
+            Some(actor_ref) => actor_ref,
+            None => return Err(ActorError::ActorNotFound),
+        };
+        drop(actors);
```

Comment:
@danielgerlag why do you use drop ? it goes out of scope and would be dropped automatically so why do it manually ?

Reply:
hmm never mind, I just realized it's a lock so you wanna release it asap. I wonder about the one on line 255 cause it's released after line 256 which is really close to it. I wonder if it's not overkill there